### PR TITLE
Reduce FS access for incremental cache

### PR DESCRIPTION
**Pull request by wyattjoh**
This passes down the route kind information to the incrememntal cache so it no longer needs to test some files existing in order to validate if the file exists or not for a route.

### CODE REVIEW

**Comment by wyattjoh** [comment]
Current dependencies on/for this PR:
* canary
  * **PR #57902** <a href="https://app.graphite.dev/github/pr/vercel/next.js/57902" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/vercel/next.js/57902?utm_source=stack-comment).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js refactor/incremental-cache-updates | Change |
| - | - | - | - |
| buildDuration | 10.2s | 10.3s | ⚠️ +119ms |
| buildDurationCached | 6s | 6.1s | N/A |
| nodeModulesSize | 175 MB | 175 MB | ⚠️ +34.1 kB |
| nextStartRea..uration (ms) | 423ms | 423ms | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js refactor/incremental-cache-updates | Change |
| - | - | - | - |
| 199-HASH.js gzip | 30 kB | 30 kB | N/A |
| 3f784ff6-HASH.js gzip | 53.2 kB | 53.2 kB | ✓ |
| 494.HASH.js gzip | 182 B | 182 B | ✓ |
| framework-HASH.js gzip | 45.5 kB | 45.5 kB | ✓ |
| main-app-HASH.js gzip | 254 B | 251 B | N/A |
| main-HASH.js gzip | 33.1 kB | 33.1 kB | N/A |
| webpack-HASH.js gzip | 1.75 kB | 1.75 kB | N/A |
| Overall change | 98.9 kB | 98.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js refactor/incremental-cache-updates | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js refactor/incremental-cache-updates | Change |
| - | - | - | - |
| _app-HASH.js gzip | 205 B | 205 B | ✓ |
| _error-HASH.js gzip | 182 B | 181 B | N/A |
| amp-HASH.js gzip | 505 B | 507 B | N/A |
| css-HASH.js gzip | 322 B | 323 B | N/A |
| dynamic-HASH.js gzip | 2.59 kB | 2.59 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 259 B | N/A |
| head-HASH.js gzip | 348 B | 347 B | N/A |
| hooks-HASH.js gzip | 369 B | 368 B | N/A |
| image-HASH.js gzip | 4.38 kB | 4.38 kB | N/A |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 318 B | 318 B | ✓ |
| script-HASH.js gzip | 384 B | 383 B | N/A |
| withRouter-HASH.js gzip | 319 B | 320 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 885 B | 885 B | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js refactor/incremental-cache-updates | Change |
| - | - | - | - |
| _buildManifest.js gzip | 484 B | 484 B | ✓ |
| Overall change | 484 B | 484 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js refactor/incremental-cache-updates | Change |
| - | - | - | - |
| index.html gzip | 528 B | 527 B | N/A |
| link.html gzip | 542 B | 544 B | N/A |
| withRouter.html gzip | 525 B | 522 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js refactor/incremental-cache-updates | Change |
| - | - | - | - |
| edge-ssr.js gzip | 96.1 kB | 96.4 kB | ⚠️ +237 B |
| page.js gzip | 140 kB | 140 kB | ⚠️ +259 B |
| Overall change | 236 kB | 237 kB | ⚠️ +496 B |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js refactor/incremental-cache-updates | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 626 B | 626 B | ✓ |
| middleware-r..fest.js gzip | 148 B | 151 B | N/A |
| middleware.js gzip | 24.8 kB | 24.8 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 2.55 kB | 2.55 kB | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: ed7d539bfd207cf324a4b3897657bc6c0b06fa4a

**Comment by ijjk** [comment]
## Failing test suites 
Commit: ed7d539bfd207cf324a4b3897657bc6c0b06fa4a

`pnpm test-dev test/development/acceptance-app/app-hmr-changes.test.ts` 
- Error overlay - RSC build errors > should handle successive HMR changes with errors correctly

<details>
<summary>Expand output</summary>

  ● Error overlay - RSC build errors › should handle successive HMR changes with errors correctly

    expect(received).toContain(expected) // indexOf

    Expected substring: "A few years ago I tweeted"
    Received string:    "<head><meta charset=\"utf-8\"><meta name=\"viewport\" content=\"width=device-width\"><meta name=\"next-head-count\" content=\"2\"><noscript data-n-css=\"\"></noscript><script defer=\"\" nomodule=\"\" src=\"/_next/static/chunks/polyfills.js?ts=1698868102635\"></script><script src=\"/_next/static/chunks/fallback/webpack.js?ts=1698868102635\" defer=\"\"></script><script src=\"/_next/static/chunks/fallback/main.js?ts=1698868102635\" defer=\"\"></script><script src=\"/_next/static/chunks/fallback/pages/_app.js?ts=1698868102635\" defer=\"\"></script><script src=\"/_next/static/chunks/fallback/pages/_error.js?ts=1698868102635\" defer=\"\"></script><noscript id=\"__next_css__DO_NOT_USE__\"></noscript></head><body style=\"overflow: hidden;\"><div id=\"__next\"></div><script src=\"/_next/static/chunks/fallback/react-refresh.js?ts=1698868102635\"></script><script id=\"__NEXT_DATA__\" type=\"application/json\">{\"props\":{\"pageProps\":{\"statusCode\":500}},\"page\":\"/_error\",\"query\":{},\"buildId\":\"development\",\"isFallback\":false,\"err\":{\"name\":\"ModuleBuildError\",\"source\":\"server\",\"message\":\"Module build failed (from ./node_modules/.pnpm/file+..+next-repo-9e75ff7ea65fc5da76bd7ebc1189729d246acbc3c44d659043a49affcd979dab+packages+n_zk23auiqadjr5fbixj5egctvwi/node_modules/@next/mdx/mdx-rs-loader.js):\\nError: page.mdx:Error: \\\"214:1: Expected a closing tag for `\\u003c/b\\u003e` (204:33) (mdx-jsx:end-tag-mismatch)\\\"\",\"stack\":\"ModuleBuildError: Module build failed (from ./node_modules/.pnpm/file+..+next-repo-9e75ff7ea65fc5da76bd7ebc1189729d246acbc3c44d659043a49affcd979dab+packages+n_zk23auiqadjr5fbixj5egctvwi/node_modules/@next/mdx/mdx-rs-loader.js):\\nError: page.mdx:Error: \\\"214:1: Expected a closing tag for `\\u003c/b\\u003e` (204:33) (mdx-jsx:end-tag-mismatch)\\\"\\n    at processResult (/tmp/next-install-8ee6191b16df8779873193b35d786fa8796fc735fa72de5b37161be3c0893890/node_modules/.pnpm/file+..+next-repo-9e75ff7ea65fc5da76bd7ebc1189729d246acbc3c44d659043a49affcd979dab+packages+n_xbeiqeitcflhwiqkpwsdc4zxhy/node_modules/next/dist/compiled/webpack/bundle5.js:28:398653)\\n    at /tmp/next-install-8ee6191b16df8779873193b35d786fa8796fc735fa72de5b37161be3c0893890/node_modules/.pnpm/file+..+next-repo-9e75ff7ea65fc5da76bd7ebc1189729d246acbc3c44d659043a49affcd979dab+packages+n_xbeiqeitcflhwiqkpwsdc4zxhy/node_modules/next/dist/compiled/webpack/bundle5.js:28:400370\\n    at /tmp/next-install-8ee6191b16df8779873193b35d786fa8796fc735fa72de5b37161be3c0893890/node_modules/.pnpm/file+..+next-repo-9e75ff7ea65fc5da76bd7ebc1189729d246acbc3c44d659043a49affcd979dab+packages+n_xbeiqeitcflhwiqkpwsdc4zxhy/node_modules/next/dist/compiled/loader-runner/LoaderRunner.js:1:8645\\n    at /tmp/next-install-8ee6191b16df8779873193b35d786fa8796fc735fa72de5b37161be3c0893890/node_modules/.pnpm/file+..+next-repo-9e75ff7ea65fc5da76bd7ebc1189729d246acbc3c44d659043a49affcd979dab+packages+n_xbeiqeitcflhwiqkpwsdc4zxhy/node_modules/next/dist/compiled/loader-runner/LoaderRunner.js:1:5828\\n    at r.callback (/tmp/next-install-8ee6191b16df8779873193b35d786fa8796fc735fa72de5b37161be3c0893890/node_modules/.pnpm/file+..+next-repo-9e75ff7ea65fc5da76bd7ebc1189729d246acbc3c44d659043a49affcd979dab+packages+n_xbeiqeitcflhwiqkpwsdc4zxhy/node_modules/next/dist/compiled/loader-runner/LoaderRunner.js:1:4039)\\n    at /tmp/next-install-8ee6191b16df8779873193b35d786fa8796fc735fa72de5b37161be3c0893890/node_modules/.pnpm/file+..+next-repo-9e75ff7ea65fc5da76bd7ebc1189729d246acbc3c44d659043a49affcd979dab+packages+n_zk23auiqadjr5fbixj5egctvwi/node_modules/@next/mdx/mdx-rs-loader.js:56:7\"},\"gip\":true,\"scriptLoader\":[]}</script><div id=\"__next-build-watcher\" style=\"position: fixed; bottom: 10px; right: 20px; width: 0px; height: 0px; z-index: 99999;\"></div><next-route-announcer><p aria-live=\"assertive\" id=\"__next-route-announcer__\" role=\"alert\" style=\"border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; top: 0px; width: 1px; white-space: nowrap; overflow-wrap: normal;\"></p></next-route-announcer><nextjs-portal></nextjs-portal></body>"

      58 |     expect(
      59 |       await session.evaluate('document.documentElement.innerHTML')
    > 60 |     ).toContain('A few years ago I tweeted')
         |       ^
      61 |
      62 |     await cleanup()
      63 |   })

      at Object.toContain (development/acceptance-app/app-hmr-changes.test.ts:60:7)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: packages/next/src/server/base-server.ts [hunk 1]
```diff
@@ -1941,14 +1946,14 @@ export default abstract class Server<ServerOptions extends Options = Options> {
...
+        // Strip the flight headers so that we generate both the RSC and HTML
+        // for the request.
         if (
           !isEdgeRuntime(opts.runtime) ||
           (this.serverOptions as any).webServerConfig
         ) {
-          for (const param of FLIGHT_PARAMETERS) {
-            delete req.headers[param.toString().toLowerCase()]
-          }
+          stripFlightHeaders(req.headers)
```
Comment (wyattjoh) [inline packages/next/src/server/base-server.ts:1956]: Also pulled this out into a helper.

File: packages/next/src/server/base-server.ts [hunk 2]
```diff
@@ -2141,141 +2148,138 @@ export default abstract class Server<ServerOptions extends Options = Options> {
...
-          renderOpts: {
-            // App Route's cannot postpone, so don't enable it.
-            ppr: false,
-            originalPathname: components.ComponentMod.originalPathname,
-            supportsDynamicHTML,
-            incrementalCache,
-            isRevalidate: isSSG,
-          },
-        }
+      if (routeModule) {
```
Comment (wyattjoh) [inline packages/next/src/server/base-server.ts:2151]: Finally I also pulled the helper methods for type detection up and exposed them.

**Review by ztanner — APPROVED** [review]

**Pull request merged**